File: app/src/main/java/com/example/sendserverdemo/RetrofitManager.java
package com.example.sendserverdemo;

import android.text.TextUtils;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitManager {

    private static volatile RetrofitManager INSTANCE;

    private static final String DEFAULT_URL = "http://192.168.1.89:8080/";

    public static RetrofitManager getInstance() {
        if (INSTANCE == null) {
            synchronized (RetrofitManager.class) {
                if (INSTANCE == null) {
                    INSTANCE = new RetrofitManager();
                }
            }
        }
        return INSTANCE;
    }

    private RetrofitManager() {
    }

    private static Retrofit retrofit;

    /**
     * 初始化Retrofit
     * @param baseUrl endIndex must "/"
     */
    public static void initRetrofit(String baseUrl) {
        HttpLoggingInterceptor httpLoggingInterceptor = new HttpLoggingInterceptor();
        httpLoggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .addInterceptor(httpLoggingInterceptor)
                .connectTimeout(5000, TimeUnit.MILLISECONDS)
                .readTimeout(5000, TimeUnit.MILLISECONDS)
                .writeTimeout(5000, TimeUnit.MILLISECONDS)
                .build();
        String url = TextUtils.isEmpty(baseUrl) ? DEFAULT_URL : baseUrl;
        retrofit = new Retrofit.Builder()
                .baseUrl(url)
                .client(okHttpClient)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
    }

    public static Retrofit getRetrofit() {
        if (retrofit != null) {
            return retrofit;
        } else {
            throw new NullPointerException("retrofit is null, please invoke initRetrofit() method before call getRetrofit()");
        }
    }

}

File: retrofit/src/main/java/com/kcst/retrofit/RetrofitManager.kt
package com.kcst.retrofit

import android.text.TextUtils
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.scalars.ScalarsConverterFactory
import java.util.concurrent.TimeUnit

class RetrofitManager private constructor() {

    companion object {
        val INSTANCE = SingleHolder.holder
        const val DEFAULT_URL="http://192.168.1.122:8080/"
    }

    private object SingleHolder {
        val holder = RetrofitManager()
    }


    lateinit var retrofit: Retrofit

    /**
     * 初始化Retrofit
     * @param baseUrl endIndex must "/"
     */
    fun initRetrofit(baseUrl: String?) {
        val httpLoggingInterceptor = HttpLoggingInterceptor()
        httpLoggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY)
        val okHttpClient: OkHttpClient = OkHttpClient.Builder()
            .addInterceptor(httpLoggingInterceptor)
            .connectTimeout(5000, TimeUnit.MILLISECONDS)
            .readTimeout(5000, TimeUnit.MILLISECONDS)
            .writeTimeout(5000, TimeUnit.MILLISECONDS)
            .build()
        val url = if (TextUtils.isEmpty(baseUrl)) DEFAULT_URL else baseUrl!!
        retrofit = Retrofit.Builder()
            .baseUrl(url)
            .client(okHttpClient)
            .addConverterFactory(ScalarsConverterFactory.create())
            .build()
    }
}